modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -2630,6 +2644,8 @@ def broadcast_apply_full_axis(
         other,
         new_index=None,
         new_columns=None,
+        new_row_lengths=None,
```

Comment:
@dchigarev's comment from #5587.

> I'm not sure about these parameters and the fact that they're present in a public method. Using them may require a deep understanding of internal partitioning logic (e.g. dealing with empty partitions) that actually shouldn't be disclosed to the outer world.
>
> My concern is that it's very easy to break everything with these parameters if you don't fully understand what's happening underneath. Here's an example of how easy it is to end up with an inconsistent frame using these parameters:

```python
import modin.pandas as pd
import modin.config as cfg
cfg.NPartitions.put(4)

df = pd.DataFrame({"a": [1, 2, 3], "b": [3, 4, 5]})
qc = df._query_compiler

def make_a_series(df):
    series = df[df.columns[0]]
    res = series.to_frame()
    assert len(res.columns) == 1
    return res

mf = qc._modin_frame.apply_full_axis(
    axis=1,
    # Completely legitimate arguments from high-level perspective
    func=make_a_series,
    new_columns=["__reduced__"],
    new_column_widths=[1],
    keep_partitioning=False
)

print(mf.column_widths) # [1], when it should be: [1, 0, 0, 0]
print(mf._partitions.shape) # (1, 4)
```

> I would propose instead the cache to be computed inside the `.broadcast_apply_full_axis`:

```python
if new_index is not None:
    if new_partitions.shape[0] == 1:
        kw["row_lengths"] = [len(new_index)]
    # TODO: add more cases where we can pre-compute partition shapes
if new_columns is not None:
    if new_partitions.shape[1] == 1:
        kw["column_widths"] = [len(new_columns)]
    # TODO: add more cases where we can pre-compute partition shapes
```

Reply:
> My concern is that it's very easy to break everything with these parameters if you don't fully understand what's happening underneath. Here's an example of how easy it is to end up with an inconsistent frame using these parameters:

Even till this moment we might have had such a case because we compute `new_row_lengths` and `new_column_widths` if `keep_partitioning=False`. `new_row_lengths` and `new_column_widths` may have zeros but this haven't bited us yet. I added a check for filtering empty partitions out in the constructor depending on the condition so we shouldn't have an issue in future.

Also, I still think that we should add `new_row_lengths` and `new_column_widths` parameters to the methods. It is useful not only for the changes in this PR, but also in https://github.com/modin-project/modin/pull/4460 and probably others.

Reply:
Major benefit of adding `new_row_lengths` and `new_column_widths` parameters is that we may not compute them manually so that we avoid more computations in the main thread.

Reply:
> I'm not sure about these parameters and the fact that they're present in a public method. Using them may require a deep understanding of internal partitioning logic (e.g. dealing with empty partitions) that actually shouldn't be disclosed to the outer world.

We need to think about the following:
1. Can we always calculate these parameters from indirect data?
2. If not, how can we protect ourselves from the misuse of these parameters? For example, can we always add missed zeroes to these parameters, based on the number of partitions? Make some additional checks, etc.

Reply:
These new parameters still look unsafe to me. I think this could be solved by the `num_splits` parameter introduced in #5585. If we would just pass `num_splits=1, new_columns=["__reduced__"]` to `broadcast_apply_full_axis` then this info should be enough to safely precompute new column widths inside `broadcast_apply_full_axis` itself.